### PR TITLE
Apply overrides before applying defaults to sidecar resource requirements

### PR DESCRIPTION
**Pull request by jeevb**
# TL;DR
Apply overrides before applying defaults to sidecar resource requirements. Related to: https://github.com/flyteorg/flyteplugins/pull/194.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Refactored code to apply resource requirements overrides from node spec first before applying default resource requirements. Found a bug in `ApplyResourceOverrides` as a result, and fixed the unit tests to properly catch the issue.

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: go/tasks/plugins/k8s/sidecar/sidecar.go [hunk 1]
```diff
@@ -39,11 +39,11 @@ func validateAndFinalizePod(
 	for index, container := range pod.Spec.Containers {
 		if container.Name == primaryContainerName {
 			hasPrimaryContainer = true
-			container.Resources = *flytek8s.ApplyResourceOverrides(ctx, container.Resources)
 			if taskCtx.TaskExecutionMetadata().GetOverrides() != nil && taskCtx.TaskExecutionMetadata().GetOverrides().GetResources() != nil {
 				resOverrides := taskCtx.TaskExecutionMetadata().GetOverrides().GetResources()
 				flytek8s.MergeResources(*resOverrides, &container.Resources)
 			}
+			container.Resources = *flytek8s.ApplyResourceOverrides(ctx, container.Resources)
```
Comment (katrogan) [inline go/tasks/plugins/k8s/sidecar/sidecar.go:46]: primary container resources should never be nil, but just in case should we check before dereferencing?

**Review by katrogan — APPROVED** [review]

File: go/tasks/plugins/k8s/sidecar/sidecar.go (hunk 1 above)
Comment (katrogan) [reply go/tasks/plugins/k8s/sidecar/sidecar.go:46]: just kidding, ApplyResourceOverrides handles initializing resources always

Comment (jeevb) [reply go/tasks/plugins/k8s/sidecar/sidecar.go:46]: Resolved offline.
